torchvision/prototype/datasets/utils/_resource.py

```diff
@@ -1,51 +1,171 @@
+    extract_archive,
+    _decompress,
+    download_file_from_google_drive,
+)
 
 
-# FIXME
-def compute_sha256(path: pathlib.Path) -> str:
-    return ""
+class OnlineResource(abc.ABC):
+    def __init__(
+        self,
+        *,
+        file_name: str,
+        sha256: Optional[str] = None,
+        decompress: bool = False,
+        extract: bool = False,
+        preprocess: Optional[Callable[[pathlib.Path], pathlib.Path]] = None,
+        loader: Optional[Callable[[pathlib.Path], IterDataPipe[Tuple[str, IO]]]] = None,
+    ) -> None:
+        self.file_name = file_name
+        self.sha256 = sha256
 
+        if preprocess and (decompress or extract):
+            warnings.warn("The parameters 'decompress' and 'extract' are ignored when 'preprocess' is passed.")
+        elif extract:
+            preprocess = self._extract
+        elif decompress:
+            preprocess = self._decompress
+        self._preprocess = preprocess
 
-class LocalResource:
-    def __init__(self, path: Union[str, pathlib.Path], *, sha256: Optional[str] = None) -> None:
-        self.path = pathlib.Path(path).expanduser().resolve()
-        self.file_name = self.path.name
-        self.sha256 = sha256 or compute_sha256(self.path)
+        if loader is None:
+            loader = self._default_loader
+        self._loader = loader
 
-    def to_datapipe(self) -> IterDataPipe:
-        return IoPathFileLoader(IterableWrapper((str(self.path),)), mode="rb")  # type: ignore
+    @staticmethod
+    def _extract(file: pathlib.Path) -> pathlib.Path:
+        return pathlib.Path(
+            extract_archive(str(file), to_path=str(file).replace("".join(file.suffixes), ""), remove_finished=False)
+        )
 
+    @staticmethod
+    def _decompress(file: pathlib.Path) -> pathlib.Path:
+        return pathlib.Path(_decompress(str(file), remove_finished=True))
 
-class OnlineResource:
-    def __init__(self, url: str, *, sha256: str, file_name: str) -> None:
-        self.url = url
-        self.sha256 = sha256
-        self.file_name = file_name
+    def _default_loader(self, path: pathlib.Path) -> IterDataPipe[Tuple[str, IO]]:
+        if path.is_dir():
+            return IoPathFileLoader(FileLister(str(path), recursive=True), mode="rb")
+
+        dp = IoPathFileLoader(IterableWrapper((str(path),)), mode="rb")
+
+        archive_loader = self._guess_archive_loader(path)
+        if archive_loader:
+            dp = archive_loader(dp)
+
+        return dp
+
+    _ARCHIVE_LOADERS = {
+        ".tar": TarArchiveReader,
+        ".zip": ZipArchiveReader,
+        ".rar": RarArchiveLoader,
+    }
+
+    def _guess_archive_loader(
+        self, path: pathlib.Path
+    ) -> Optional[Callable[[IterDataPipe[Tuple[str, IO]]], IterDataPipe[Tuple[str, IO]]]]:
+        try:
+            _, archive_type, _ = _detect_file_type(path.name)
+        except RuntimeError:
+            return None
+        return self._ARCHIVE_LOADERS.get(archive_type)  # type: ignore[arg-type]
```

Comment:
See facebookexternal/torchdata#23. cc @ejguan

Reply:
IMHO, it's different. One option as you have pointed out is to have a class to read files from all different archives.
 The another option is to return an archiveReader for a specific archive type.
From my perspective, the latter solution is more suitable because the first solution needs to rely on that all file handles opened from tar, zip and rar would behave same. Otherwise, it's gonna be hard to debug which file handles cause a problem. And, for the first solution, such DataPipe class would need more dependencies like rarfile. If users only need to read from tar or zip, by using such DataPipe, they have to install rarfile.

Reply:
> rely on that all file handles opened from tar, zip and rar would behave same.

I'm not sure I understand. If we had an `ArchiveReader` datapipe, in its `__iter__` method it could infer the datapipe format from the path and than use whatever functionality it needs to iterate the archive. Something along the lines of 

```python
def iterate_tar(path, file):
    ...

def iterate_zip(path, file):
    ...

class ArchiveReader(IterDataPipe):
    ...

    def infer_archive_type(self, path):
        ...

    def __iter__(self):
        for path, file in self.datapipe:
            archive_type = self.infer_archive_type(path)
            if archive_type == "tar":
                yield from iterate_tar(path, file)
            elif archive_type == "zip":
                yield from iterate_zip(path, file)
            else:
                raise ValueError("Unknown archive type!")

class TarReader(IterDataPipe):
    ...

    def __iter__(self):
        for path, file in self.datapipe:
            yield from iterate_tar(path, file)
            
class ZipReader(IterDataPipe):
    ...

    def __iter__(self):
        for path, file in self.datapipe:
            yield from iterate_zip(path, file)
```

If the user wants full control, they could still use a specific archive reader. Otherwise they can use the convenience of not needing to specify the archive type by using `ArchiveReader`.

> And, for the first solution, such DataPipe class would need more dependencies like rarfile. If users only need to read from tar or zip, by using such DataPipe, they have to install rarfile.

The dependencies could be checked at runtime. To stick with the `rarfile` example, we could lazily import it in the `archive_type == "rar"` branch. Meaning, if the user never hits a rar archive, everything works fine.

Reply:
> rely on that all file handles opened from tar, zip and rar would behave same.

I am thinking if there is difference between the behavior of file handles yielded from such DataPipe. If so, all file handles from a single DataPipe can cause confusion.

> The dependencies could be checked at runtime. To stick with the rarfile example, we could lazily import it in the archive_type == "rar" branch. Meaning, if the user never hits a rar archive, everything works fine.

But, then we can not raise Error at construction time. The iteration of pipeline would stop in the middle of training.

The main benefit for the first solution would be single DataPipe handles all types of archives. This would be useful only if domain has the use case that a single Dataset owns different archives.
But, the latter solution is same as what you are currently doing to return a specific archive reader per dataset.

Reply:
> I am thinking if there is difference between the behavior of file handles yielded from such DataPipe. If so, all file handles from a single DataPipe can cause confusion.

IMO when a user opts to use a datapipe like `ArchiveReader`, they enter a contract that says "I'm not going to do anything with the file handles that is specific to the archive type". They all behave the same for simply reading the data, right?